prototype_source/gpu_quantization_torchao_tutorial.py

```diff
@@ -0,0 +1,286 @@
+# bf16 runtime of the block is 25.43ms and peak memory  3.17GB
+
+
+######################################################################
+# Just this quick change improves runtime by a factor of ~7x in my
+# experience (186.16ms to 25.43ms).
+#
+# Next we torch.compile the model and see how much the performance
+# improves
+#
+
+model_c = torch.compile(model, mode='max-autotune')
+comp_res = benchmark(model_c, image)
+print(f"bf16 compiled runtime of the block is {comp_res['time']:0.2f}ms and peak memory {comp_res['memory']: 0.2f}GB")
+# bf16 compiled runtime of the block is 19.95ms and peak memory  2.24GB
+
+
+######################################################################
+# The first time this is run, you should see a sequence of AUTOTUNE
+# outputs which occurs when inductor compares the performance between
+# various kernel parameters for a kernel. This only happens once (unless
+# you delete your cache) so if you run the cell again you should just get
+# the benchmark output.
+#
+# torch.compile yields about another 27% improvement. This brings the
+# model to a reasonable baseline where we now have to work a bit harder
+# for improvements.
+#
+# Now lets apply quantization. Quantization for GPUs comes in 3 main forms
+# in `torchao <https://github.com/pytorch-labs/ao>`_ which is just native
+# pytorch+python code. We have int8 dynamic quantization, int8 weight-only
+# quantization and int4 weight-only quantization. Different models, or
+# sometimes different layers in a model can require different techniques.
+# For models which are heavily compute bound, dynamic quantization tends
+# to work the best since it swaps the normal expensive floating point
+# matmul ops with integer versions. Weight-only quantization works better
+# in memory bound situations where the benefit comes from loading less
+# weight data, rather than doing less computation. The torchao api’s:
+#
+# ``change_linear_weights_to_int8_dqtensors``,
+# ``change_linear_weights_to_int8_woqtensors`` or
+# ``change_linear_weights_to_int4_woqtensors``
+#
+# can be used to easily apply the desired quantization technique and then
+# once the model is torch.compile-d with max-autotune, quantization is
+# complete and we can see our speedup.
+#
+# Note: These api’s may be buggy on older versions of pytorch, if you run
+# into that issue, you can instead use
+#
+# ``apply_dynamic_quant`` and ``apply_weight_only_int8_quant``
+#
+# As drop in replacement for the 2 above (no replacement for int4). The
+# difference between the two apis is that ‘change_linear_weights’ api’s
+# alter the weight tensor of the linear module so instead of doing a
+# normal linear, it does a quantized operation. This is helpful when users
+# have non-standard linear ops that do more than 1 thing. The ‘apply’
+# api’s directly swap the linear modules for a quantized module which
+# works on older versions but doesn’t work with non-standard linear
+# modules.
```

Comment:
```suggestion
#  The difference between the two APIs is that ``change_linear_weights`` API
# alters the weight tensor of the linear module so instead of doing a
# normal linear, it does a quantized operation. This is helpful when you
# have non-standard linear ops that do more than one thing. The ``apply``
# APIs directly swap the linear modules for a quantized module which
# works on older versions but doesn’t work with non-standard linear
# modules.
```